### PR TITLE
chore: pipelines via user-data COMPASS-7186

### PR DESCRIPTION
**Pull request by mabaasit**
### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-aggregations/src/utils/pipeline-storage.ts [hunk 1]
```diff
@@ -1,26 +1,10 @@
-import { promises as fs } from 'fs';
-import path from 'path';
+import { z } from 'zod';
+import { UserData } from '@mongodb-js/compass-user-data';
```
Comment (gribnoysup) [inline packages/compass-aggregations/src/utils/pipeline-storage.ts:2]: Have you considered re-exporting zod from user-data or integrating it into class constructor? You always need zod when using user-data, every package has x2 dependencies and need to maintain the version parity now, with re-export it should be easier to maintain

File: packages/compass-aggregations/src/utils/pipeline-storage.ts [hunk 2]
```diff
@@ -57,88 +41,84 @@ function savedPipelineToText(pipeline: StoredPipeline['pipeline']): string {
...
       ...attributes,
-    };
-    await fs.writeFile(
-      filePath,
-      JSON.stringify(updated, null, 2),
-      ENCODING_UTF8
-    );
-    return updated;
+    });
+    return await this.userData.readOne(id);
```
Comment (gribnoysup) [inline packages/compass-aggregations/src/utils/pipeline-storage.ts:118]: We already have this document here, do we need to read it again?

File: packages/compass-aggregations/src/utils/pipeline-storage.ts [hunk 3]
```diff
@@ -57,88 +41,84 @@ function savedPipelineToText(pipeline: StoredPipeline['pipeline']): string {
...
-      ).filter(Boolean) as StoredPipeline[];
+      const { data } = await this.userData.readAll({
+        ignoreErrors: false,
+        readFileStats: true,
+        mergeStats(input, stats) {
+          return {
+            ...input,
+            lastModified: Number(stats.mtimeMs),
+          };
+        },
```
Comment (gribnoysup) [inline packages/compass-aggregations/src/utils/pipeline-storage.ts:103]: Stats seem to be merged for load all but not for load one, is this intentional change?

File: packages/compass-user-data/src/user-data.ts [hunk 4]
```diff
@@ -7,26 +8,35 @@ import type { z } from 'zod';
...
   ignoreErrors: boolean;
 };
 
 export class UserData<T extends z.Schema> {
   private readonly subdir: string;
   private readonly basePath?: string;
-  private readonly serialize: SerializeContent<T>;
-  private readonly deserialize: DeserializeContent;
+  private readonly serialize: SerializeContent<z.input<T>>;
+  private readonly deserialize: DeserializeContent<z.input<T>>;
```
Comment (gribnoysup) [inline packages/compass-user-data/src/user-data.ts:39]: That's not true, at the moment this is called, nothing guarantees that the value returned is matching T

File: packages/compass-aggregations/src/utils/pipeline-storage.ts (hunk 3 above)
Comment (mabaasit) [reply packages/compass-aggregations/src/utils/pipeline-storage.ts:103]: nice catch! missed it completely.

File: packages/compass-aggregations/src/utils/pipeline-storage.ts (hunk 2 above)
Comment (mabaasit) [reply packages/compass-aggregations/src/utils/pipeline-storage.ts]: The reason we are doing another find here it to get the latest data which is transformed by the schema validator to match the output.

We did talk about it last week and agreed on having this encapsulated within `userdata.write` to return us the latest data. However the userdata does not know the exact structure of the data [can be array (shell-history), object (almost everywhere) or possibly a scaler value] and merging the current value with the update attributes varies and for that reason i'll leave that to the user to handle it.

File: packages/compass-aggregations/src/utils/pipeline-storage.ts (hunk 1 above)
Comment (mabaasit) [reply packages/compass-aggregations/src/utils/pipeline-storage.ts]: added in [f153397](https://github.com/mongodb-js/compass/pull/4797/commits/f153397ef1e4b8c6d8244fa0b5ab6331991cfae2)

File: packages/compass-aggregations/src/utils/pipeline-storage.ts (hunk 2 above)
Comment (gribnoysup) [reply packages/compass-aggregations/src/utils/pipeline-storage.ts]: I still think we need to do something for this, we are sacrificing the most common use-case here for one corner case. Feel free to leave it as-is though, I might have time to look into it when tackling switching atlas-service to using user data

File: packages/compass-aggregations/src/utils/pipeline-storage.ts [hunk 5]
```diff
@@ -57,88 +41,109 @@ function savedPipelineToText(pipeline: StoredPipeline['pipeline']): string {
...
-    }
+  async createOrUpdate(id: string, attributes: SaveablePipeline) {
+    const pipelineExists = Boolean(
+      await this.userData.readOne(id, {
+        ignoreErrors: true,
+      })
+    );
+    return await (pipelineExists
+      ? this.updateAttributes(id, attributes)
+      : this.create(attributes));
```
Comment (gribnoysup) [inline packages/compass-aggregations/src/utils/pipeline-storage.ts:126]: We're adding another read on top of all the reads that `update` will be doing, I think we need to reconsider some thing about internal workings of user data based on the use-cases we have in compass, doesn't look like the way it works fits well to what we need to do most of the time

**Review by gribnoysup — APPROVED** [review]
I think it's good to be merged for now, but the way user data is used here shows that the interface we created just doesn't allow for easy use inside the application so we probably need to re-asses how it works at some point

**Pull request merged**